packages/osd-agents/src/agents/langgraph/prompt_manager.ts

```diff
@@ -0,0 +1,335 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { readFileSync, existsSync } from 'fs';
+import { join } from 'path';
+import * as yaml from 'js-yaml';
+import { Logger } from '../../utils/logger';
+import { BaseMCPClient } from '../../mcp';
+
+export class PromptManager {
+  private logger: Logger;
+  private mcpClients: Record<string, BaseMCPClient>;
+  private baseSystemPrompt: string = '';
+
+  constructor(logger: Logger, mcpClients: Record<string, BaseMCPClient>) {
+    this.logger = logger;
+    this.mcpClients = mcpClients;
+  }
+
+  /**
+   * Load and enhance the system prompt with dynamic MCP tool information
+   */
+  loadSystemPrompt(customSystemPrompt?: string): void {
+    if (customSystemPrompt) {
+      this.baseSystemPrompt = this.enhanceSystemPrompt(customSystemPrompt);
+      this.logger.info('Using enhanced custom system prompt with dynamic content', {
+        customPromptLength: customSystemPrompt.length,
+        finalPromptLength: this.baseSystemPrompt.length,
+        customPromptPreview: customSystemPrompt.substring(0, 200) + '...',
+      });
+    } else {
+      // Always use dynamic system prompt that describes actual MCP tools
+      this.baseSystemPrompt = this.getDefaultSystemPrompt();
+      this.logger.info('Using dynamic system prompt with MCP tools', {
+        promptLength: this.baseSystemPrompt.length,
+        connectedServers: Object.keys(this.mcpClients).length,
+        promptPreview: this.baseSystemPrompt.substring(0, 200) + '...',
+      });
+    }
+  }
+
+  /**
+   * Get the base system prompt (without client-specific data)
+   */
+  getBaseSystemPrompt(): string {
+    return this.baseSystemPrompt;
+  }
+
+  /**
+   * Inject client-specific data into the system prompt
+   */
+  injectClientDataIntoPrompt(
+    clientState?: any,
+    clientContext?: any[],
+    clientTools?: any[]
+  ): string {
+    let prompt = this.baseSystemPrompt;
+
+    // Inject client state with helpful usage instructions
+    if (clientState) {
+      const stateContent = `
+## Current Session State
+The following state is maintained and synchronized with the client:
+
+\`\`\`json
+${JSON.stringify(clientState, null, 2)}
+\`\`\`
+
+### How to Use State:
+- This state persists across the conversation
+- You can modify state values to track progress or store information
+- State changes will be synchronized with the client via STATE_DELTA events
+- Use state to maintain context between tool calls and responses
+- Example: Track investigation steps, store intermediate results, maintain counters
+- **Dataset**: If \`dataContext.dataset.title\` exists, use it as the index pattern for OpenSearch queries
+- **Query**: The \`query\` field contains the current PPL query shown in the UI - modify it when users request query changes
+`;
+      prompt = prompt.replace('{{CLIENT_STATE}}', stateContent);
+    } else {
+      prompt = prompt.replace('{{CLIENT_STATE}}', '// No client state provided');
```

Comment:
what is `{{CLIENT_STATE}}`? can this placeholder come from some constant?

would `clientState` ever be a complex/circular object that can't be stringified?